recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,425 @@
+  basicAuthentication
+
+  @doc("azureWorkloadIdentity type is used to represent registry authentication using azure federated identity and the secretstore resource is expected to have the keys 'clientId' and 'tenantId'")
+  azureWorkloadIdentity
+
+  @doc("awsIRSA type is used to represent registry authentication using AWS IRSA(IAM Roles for Service accounts) and the secretstore resource is expected to have the keys 'roleARN'")
+  awsIRSA
+}
+
+@doc("Registry Secret Configuration used to authenticate to private bicep registries.")
+model RegistrySecretConfig {
+  @doc("The ID of an Applications.Core/SecretStore resource containing credential information used to authenticate private container registry.The keys in the secretstore depends on the type.")
+  secretSource?: string;
+
+  @doc(RegistryAuthType represents the authentication type used to authenticate private Bicep registries.)
+  type: RegistryAuthType
+}
+```
+Pros:
+- Works well with current recipe config secrets design.
+
+Cons:
+- Adding a new property to specify the `type` of container registry authentication.
+- Incorrect configurations (e.g: clientId is not provided as a secret key for type azureWorkloadIdentity) are not known until recipe deployment.
+
+Option 2:
+
+Adding support to the `Applications.Core/secretStores` resource to include additional types. Today `secretStores` resource only supports types `certificate` and `generic`.
+
+secretstores.tsp
+
+```diff
+@doc("The type of SecretStore data")
+enum SecretStoreDataType {
+  @doc("Generic secret data type")
+  generic,
+
+  @doc("Certificate secret data type")
+  certificate,
+
++  @doc("basicAuthentication type is used to represent username and password based authentication and the secretstore resource is expected to have the keys 'username' and 'password'")
++  basicAuthentication
++
++  @doc("azureWorkloadIdentity type is used to represent registry authentication using azure federated identity and the secretstore resource is expected to have the keys 'clientId' and 'tenantId'")
++  azureWorkloadIdentity
++
++  @doc("awsIRSA type is used to represent registry authentication using AWS IRSA.")
++  awsIRSA
+}
+```
+Pros:
+- Leveraging the existing property `type` in `Application.core/secretStores` to specify the kind of authentication.
+- Incorrect configurations(e.g: clientId is not provided as a secret key for type azureWorkloadIdentity) can be detected during the creation of the `Applications.Core/secretStore`.
+
+Cons:
+- This options lead to either having driver specific code in engine/secretLoader or driver an making api call to corerp to get secrets information.
+
+I propose adding `type` property to the `RegistrySecretConfig` i.e Option 1, as with the current implementation of the secrets flow between engine-driver,  Option 2 lead to either having driver specific code in engine/secretLoader or driver an making api call to corerp to get secrets information. 
+
```

Comment:
I don't think that either option is bad, but I'm curious to hear what was discussed last week.